src/zope/password/password.py

```diff
@@ -582,9 +571,105 @@ def match(self, hashed_password):
+    """
+    BCRYPT KDF password manager.
+
+    This manager converts a plain text password into a byte array .
+    The password and salt values (randomly generated when the password
+    is encoded) are combined and repeatedly hashed *rounds* times. The
+    repeated hashing is designed to thwart discovery of the key via
+    password guessing attacks. The higher the number of rounds, the
+    slower each attempt will be.
+
+    Compared to the :class:`BCRYPTPasswordManager`, this has the
+    advantage of allowing tunable rounds, so as computing devices get
+    more powerful making brute force attacks faster, the difficulty
+    level can be raised (for newly encoded passwords).
+
+    >>> from zope.password.password import BCRYPTKDFPasswordManager
+    >>> manager = BCRYPTKDFPasswordManager()
+    >>> manager.checkPassword(b'not from here', None)
+    False
+
+    Let's encode a password. We'll use the minimum acceptable number
+    of rounds so that the tests run fast:
+
+    >>> manager.rounds = 51
+    >>> password = u"right \N{CYRILLIC CAPITAL LETTER A}"
+    >>> encoded = manager.encodePassword(password)
+    >>> print(encoded.decode())
+    {BCRYPTKDF}33...
+
+    It checks out:
+
+    >>> manager.checkPassword(encoded, password)
+    True
+
+    We can change the number of rounds for future encodings:
+
+    >>> manager.rounds = 100
+    >>> encoded2 = manager.encodePassword(password)
+    >>> print(encoded2.decode())
+    {BCRYPTKDF}64...
+    >>> manager.checkPassword(encoded2, password)
+    True
+
+    And the old password still checks out:
+
+    >>> manager.checkPassword(encoded, password)
+    True
+    """
+
+    #: The number of rounds of hashing that should be applied.
+    #: The higher the number, the slower it is. It should be at least
+    #: 50.
+    rounds = 1<<10
+
+    #: The number of bytes long the encoded password will be. It must be
+    #: at least 1 and no more than 512.
+    keylen = 32
+
+    _prefix = b'{BCRYPTKDF}'
+
+    def _encode(self, password, salt, rounds, keylen):
+        password = _encoder(password)
+
+        key = bcrypt.kdf(password, salt=salt,
+                         desired_key_bytes=keylen,
+                         rounds=rounds)
+        rounds_bytes = _encoder('%x' % rounds)
+        result = (self._prefix
+                  + rounds_bytes
+                  + b'$'
+                  + urlsafe_b64encode(salt)
+                  + b'$'
+                  + urlsafe_b64encode(key))
+        return result
+
+    def encodePassword(self, password):
+        salt = bcrypt.gensalt()
+        return self._encode(password, salt, self.rounds, self.keylen)
+
+    def checkPassword(self, hashed_password, clear_password):
+        hashed_password = _encoder(hashed_password)
+        if not self.match(hashed_password):
+            return False
+        rounds, salt, key = hashed_password[len(self._prefix):].split(b'$')
+        rounds = int(rounds, 16)
```

Comment:
I'm worrying that it's easy to make this function crash if you provide corrupt hashes (e.g. wrong number of '$' signs or non-hex digits in the `rounds` portion).

I am not a crypto person and I don't know whether that's fine (hashes are probably not attacker-supplied?) or not.

Reply:
We want to avoid false positives here, so a crash is better than that, I think. But the hashed password would not be supplied by the attacher, the `clear_password` is. The hashed password is stored by the system and previously generated by this object, if it's being used according to the contract of `IMatchingPasswordManager`.